services/venice-router/src/main/java/com/linkedin/venice/router/MetaDataHandler.java

```diff
@@ -158,63 +156,67 @@ public void channelRead0(ChannelHandlerContext ctx, HttpRequest req) throws IOEx
-        ReferenceCountUtil.retain(req);
-        ctx.fireChannelRead(req);
+    try {
+      switch (resourceType) {
+        case TYPE_LEADER_CONTROLLER:
+        case TYPE_LEADER_CONTROLLER_LEGACY:
+          // URI: /leader_controller or /master_controller
+          handleControllerLookup(ctx);
+          break;
+        case TYPE_KEY_SCHEMA:
+          // URI: /key_schema/${storeName}
+          // For key schema lookup, we only consider storeName
+          handleKeySchemaLookup(ctx, helper);
+          break;
+        case TYPE_VALUE_SCHEMA:
+          // The request could fetch one value schema by id or all the value schema for the given store
+          // URI: /value_schema/{$storeName} - Get all the value schema
+          // URI: /value_schema/{$storeName}/{$valueSchemaId} - Get single value schema
+          handleValueSchemaLookup(ctx, helper);
+          break;
+        case TYPE_LATEST_VALUE_SCHEMA:
+          // The request could fetch the latest value schema for the given store
+          // URI: /latest_value_schema/{$storeName} - Get the latest value schema
+          handleLatestValueSchemaLookup(ctx, helper);
+          break;
+        case TYPE_GET_UPDATE_SCHEMA:
+          // URI: /update_schema/{$storeName} - Get all the update schema
+          // URI: /update_schema/{$storeName}/{$valueSchemaId} - Get single update schema
+          // The request could fetch the latest derived update schema of a specific value schema
+          handleUpdateSchemaLookup(ctx, helper);
+          break;
+        case TYPE_CLUSTER_DISCOVERY:
+          // URI: /discover_cluster/${storeName}
+          handleD2ServiceLookup(ctx, helper, req);
+          break;
+        case TYPE_RESOURCE_STATE:
+          // URI: /resource_state
+          handleResourceStateLookup(ctx, helper);
+          break;
+        case TYPE_PUSH_STATUS:
+          // URI: /push_status
+          handlePushStatusLookUp(ctx, helper);
+          break;
+        case TYPE_STREAM_HYBRID_STORE_QUOTA:
+          handleStreamHybridStoreQuotaStatusLookup(ctx, helper);
+          break;
+        case TYPE_STREAM_REPROCESSING_HYBRID_STORE_QUOTA:
+          handleStreamReprocessingHybridStoreQuotaStatusLookup(ctx, helper);
+          break;
+        case TYPE_STORE_STATE:
+          handleStoreStateLookup(ctx, helper);
+          break;
+        case TYPE_REQUEST_TOPIC:
+          handleRequestTopic(ctx, helper, req);
+          break;
+        default:
+          // SimpleChannelInboundHandler automatically releases the request after channelRead0 is done.
+          // since we're passing it on to the next handler, we need to retain an extra reference.
+          ReferenceCountUtil.retain(req);
+          ctx.fireChannelRead(req);
+      }
+    } catch (VeniceHttpException e) {
+      setupResponseAndFlush(HttpResponseStatus.valueOf(e.getHttpStatusCode()), e.getMessage().getBytes(), false, ctx);
```

Comment:
My understanding is you add this try catch here. Just wondering what would happen in the past if handler throws exception in Router request?

Reply:
Previously, router framework would send a 500 error to client with the exception message. But here, we set a proper status code in the response